packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -126,96 +172,242 @@ contract MixinExchangeCore is
+        fillResults.takerFeePaid = getPartialAmount(
+            fillResults.takerAssetFilledAmount,
+            order.takerAssetAmount,
+            order.takerFee);
+
+        status = uint8(Status.SUCCESS);
+        return;
+    }
+
+    /// @dev Updates state with results of a fill order.
+    /// @param order that was filled.
+    /// @param takerAddress Address of taker who filled the order.
+    /// @return fillResults Amounts filled and fees paid by maker and taker.
+    function updateFilledState(
+        Order memory order,
+        address takerAddress,
+        bytes32 orderHash,
+        FillResults memory fillResults)
+        internal
+    {
         // Update state
         filled[orderHash] = safeAdd(filled[orderHash], fillResults.takerAssetFilledAmount);
 
-        // Settle order
-        (fillResults.makerAssetFilledAmount, fillResults.makerFeePaid, fillResults.takerFeePaid) =
-            settleOrder(order, takerAddress, fillResults.takerAssetFilledAmount);
-
         // Log order
         emitFillEvent(order, takerAddress, orderHash, fillResults);
-        return fillResults;
     }
 
-    /// @dev After calling, the order can not be filled anymore.
+    /// @dev Fills the input order.
     /// @param order Order struct containing order specifications.
-    /// @return True if the order state changed to cancelled.
-    ///         False if the transaction was already cancelled or expired.
-    function cancelOrder(Order memory order)
+    /// @param takerAssetFillAmount Desired amount of takerToken to sell.
+    /// @param signature Proof that order has been created by maker.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function fillOrder(
+        Order memory order,
+        uint256 takerAssetFillAmount,
+        bytes memory signature)
         public
-        returns (bool)
+        returns (FillResults memory fillResults)
     {
-        // Compute the order hash
-        bytes32 orderHash = getOrderHash(order);
+        // Fetch order info
+        bytes32 orderHash;
+        uint8 orderStatus;
+        uint256 takerAssetFilledAmount;
+        (orderStatus, orderHash, takerAssetFilledAmount) = getOrderInfo(order);
+
+        // Fetch taker address
+        address takerAddress = getCurrentContextAddress();
+
+        // Either our context is valid or we revert
+        validateFillOrderContextOrRevert(order, orderStatus, orderHash, takerAssetFilledAmount, signature, takerAddress, takerAssetFillAmount);
+
+        // Compute proportional fill amounts
+        uint8 status;
+        (status, fillResults) = calculateFillResults(order, orderStatus, takerAssetFilledAmount, takerAssetFillAmount);
+        if (status != uint8(Status.SUCCESS)) {
+            emit ExchangeStatus(uint8(status), orderHash);
+            return fillResults;
+        }
+
+        // Settle order
+        (fillResults.makerAssetFilledAmount, fillResults.makerFeePaid, fillResults.takerFeePaid) =
```

Comment:
Isn't `fillResults` already populated from `calculateFillResults`? We can just make `settleOrder` take the `fillResults` instance and not do the calculations again within `settleOrder`.